File: app/src/main/java/gm/tieba/tabswitch/dao/Preferences.java
package gm.tieba.tabswitch.dao;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class Preferences {
    private static SharedPreferences sTsPreferences;
    private static SharedPreferences sTsConfig;
    private static SharedPreferences sTsNotes;

    public static void init(final Context context) {
        sTsPreferences = context.getSharedPreferences("TS_preferences", Context.MODE_PRIVATE);
        sTsConfig = context.getSharedPreferences("TS_config", Context.MODE_PRIVATE);
        sTsNotes = context.getSharedPreferences("TS_notes", Context.MODE_PRIVATE);
    }

    // Preferences
    public static Map<String, ?> getAll() {
        return sTsPreferences.getAll();
    }

    public static void remove(final String key) {
        final SharedPreferences.Editor editor = sTsPreferences.edit();
        editor.remove(key);
        editor.apply();
    }

    public static void putBoolean(final String key, final boolean value) {
        final SharedPreferences.Editor editor = sTsPreferences.edit();
        editor.putBoolean(key, value);
        editor.apply();
    }

    public static boolean getBoolean(final String key) {
        return sTsPreferences.getBoolean(key, false);
    }

    public static void putString(final String key, final String value) {
        final SharedPreferences.Editor editor = sTsPreferences.edit();
        editor.putString(key, value);
        editor.apply();
    }

    public static String getString(final String key) {
        return sTsPreferences.getString(key, null);
    }

    public static void putStringSet(final String key, final String value, final boolean isContain) {
        final List<String> list = new ArrayList<>(getStringSet(key));
        if (!isContain) list.remove(value);
        else if (!list.contains(value)) list.add(value);

        final SharedPreferences.Editor editor = sTsPreferences.edit();
        editor.putStringSet(key, new HashSet<>(list));
        editor.apply();
    }

    public static Set<String> getStringSet(final String key) {
        return sTsPreferences.getStringSet(key, new HashSet<>());
    }

    // Config
    public static void putEULAAccepted() {
        final SharedPreferences.Editor editor = sTsConfig.edit();
        editor.putBoolean("EULA", true);
        editor.apply();
    }

    public static boolean getIsEULAAccepted() {
        return sTsConfig.getBoolean("EULA", false);
    }

    public static void putAutoSignEnabled() {
        final SharedPreferences.Editor editor = sTsConfig.edit();
        editor.putBoolean("auto_sign", true);
        editor.apply();
    }

    public static boolean getIsAutoSignEnabled() {
        return sTsConfig.getBoolean("auto_sign", false);
    }

    @SuppressLint("ApplySharedPref")
    public static void putPurgeEnabled() {
        final SharedPreferences.Editor editor = sTsConfig.edit();
        editor.putBoolean("ze", true);
        editor.commit();
    }

    public static boolean getIsPurgeEnabled() {
        return sTsConfig.getBoolean("ze", false);
    }

    @SuppressLint("ApplySharedPref")
    public static void putSignature(final int i) {
        final SharedPreferences.Editor editor = sTsConfig.edit();
        editor.putInt("signature", i);
        editor.commit();
    }

    public static int getSignature() {
        return sTsConfig.getInt("signature", 0);
    }

    public static void putLikeForum(final Set<String> follow) {
        final SharedPreferences.Editor editor = sTsConfig.edit();
        editor.putStringSet("like_forum", follow);
        editor.apply();
    }

    public static Set<String> getLikeForum() {
        return sTsConfig.getStringSet("like_forum", null);
    }

    public static void putSignDate() {
        final SharedPreferences.Editor editor = sTsConfig.edit();
        editor.putInt("sign_date", Calendar.getInstance().get(Calendar.DAY_OF_YEAR));
        editor.apply();
    }

    public static boolean getIsSigned() {
        return Calendar.getInstance().get(Calendar.DAY_OF_YEAR) == sTsConfig.getInt("sign_date", 0);
    }

    // Notes
    public static Map<String, ?> getNotes() {
        return sTsNotes.getAll();
    }

    public static String getNote(final String name) {
        return sTsNotes.getString(name, null);
    }

    public static SharedPreferences.Editor getTsNotesEditor() {
        return sTsNotes.edit();
    }
}

File: app/src/main/java/gm/tieba/tabswitch/dao/Preferences.kt
package gm.tieba.tabswitch.dao

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import java.util.Calendar

object Preferences {
    private lateinit var sTsPreferences: SharedPreferences
    private lateinit var sTsConfig: SharedPreferences

    fun init(context: Context) {
        sTsPreferences = context.getSharedPreferences("TS_preferences", Context.MODE_PRIVATE)
        sTsConfig = context.getSharedPreferences("TS_config", Context.MODE_PRIVATE)
    }

    fun getAll(): Map<String, *> = sTsPreferences.all

    fun remove(key: String) {
        sTsPreferences.edit().remove(key).apply()
    }

    fun putBoolean(key: String, value: Boolean) {
        sTsPreferences.edit().putBoolean(key, value).apply()
    }

    @JvmStatic
    fun getBoolean(key: String): Boolean = sTsPreferences.getBoolean(key, false)

    fun putString(key: String, value: String) {
        sTsPreferences.edit().putString(key, value).apply()
    }

    fun getString(key: String): String? = sTsPreferences.getString(key, null)

    fun putStringSet(key: String, value: String, isContain: Boolean) {
        val set = getStringSet(key).toMutableSet()
        if (isContain) {
            set.add(value)
        } else {
            set.remove(value)
        }
        sTsPreferences.edit().putStringSet(key, set).apply()
    }

    fun getStringSet(key: String): Set<String> = sTsPreferences.getStringSet(key, emptySet()) ?: emptySet()

    // Config
    fun putEULAAccepted() {
        sTsConfig.edit().putBoolean("EULA", true).apply()
    }

    fun getIsEULAAccepted(): Boolean = sTsConfig.getBoolean("EULA", false)

    fun putAutoSignEnabled() {
        sTsConfig.edit().putBoolean("auto_sign", true).apply()
    }

    fun getIsAutoSignEnabled(): Boolean = sTsConfig.getBoolean("auto_sign", false)

    fun getTransitionAnimationEnabled(): Boolean =
        sTsPreferences.getBoolean("transition_animation", false)

    @SuppressLint("ApplySharedPref")
    fun putPurgeEnabled() {
        sTsConfig.edit().putBoolean("ze", true).commit()
    }

    fun getIsPurgeEnabled(): Boolean = sTsConfig.getBoolean("ze", false)

    @SuppressLint("ApplySharedPref")
    fun putSignature(i: Int) {
        sTsConfig.edit().putInt("signature", i).commit()
    }

    fun getSignature(): Int = sTsConfig.getInt("signature", 0)

    fun putLikeForum(follow: Set<String?>?) {
        sTsConfig.edit().putStringSet("like_forum", follow).apply()
    }

    fun getLikeForum(): Set<String>? = sTsConfig.getStringSet("like_forum", null)

    fun putSignDate() {
        sTsConfig.edit().putInt("sign_date", Calendar.getInstance()[Calendar.DAY_OF_YEAR]).apply()
    }

    fun getIsSigned(): Boolean =
        Calendar.getInstance()[Calendar.DAY_OF_YEAR] == sTsConfig.getInt("sign_date", 0)

    @SuppressLint("ApplySharedPref")
    fun commit() {
        sTsConfig.edit().commit()
        sTsPreferences.edit().commit()
    }
}
